src/libDirectoryService/DSBlockPreProcessing.cpp

```diff
@@ -564,6 +527,52 @@ bool DirectoryService::VerifyPoWOrdering(
   return ret;
 }
 
+bool DirectoryService::VerifyPoWFromLeader(const Peer& peer,
+                                           const PubKey& pubKey,
+                                           const PoWSolution& powSoln) {
+  auto headerHash = POW::GenHeaderHash(
+      m_mediator.m_dsBlockRand, m_mediator.m_txBlockRand, peer.m_ipAddress,
+      pubKey, powSoln.lookupId, powSoln.gasPrice);
+
+  auto difficulty =
+      (GUARD_MODE && Guard::GetInstance().IsNodeInShardGuardList(pubKey))
+          ? (POW_DIFFICULTY / POW_DIFFICULTY)
+          : m_mediator.m_dsBlockChain.GetLastBlock()
+                .GetHeader()
+                .GetDifficulty();
+
+  string resultStr, mixHashStr;
+  if (!DataConversion::charArrToHexStr(powSoln.result, resultStr)) {
+    return false;
+  }
+
+  if (!DataConversion::charArrToHexStr(powSoln.mixhash, mixHashStr)) {
+    return false;
+  }
+
+  if (!POW::GetInstance().PoWVerify(m_pendingDSBlock->GetHeader().GetBlockNum(),
+                                    difficulty, headerHash, powSoln.nonce,
+                                    resultStr, mixHashStr)) {
+    LOG_GENERAL(WARNING, "Failed to verify PoW solution from leader for node: "
+                             << pubKey);
+    return false;
+  }
+
+  m_allPoWs[pubKey] = powSoln;
```

Comment:
let's use at() ?

Reply:
This is an insert operation, i will use insert instead

Reply:
This is an insert operation, cannot use at